File: Version-JavaScript/models/Order.js
import { v4 as uuidv4 } from 'uuid';

// Implementation of class and methods for placing and managing customer orders
class Order {
    constructor(customerId, productId, quantity) {
        if (!customerId || !productId || !quantity) {
            throw new Error("All required fields must be provided!");
        }

        this.id = uuidv4();
        this.customerId = customerId;
        this.productId = productId;
        this.quantity = quantity;
        this.date = new Date();
    }
    
    //! Method to Add a new product to the inventory.
	createOrder(supermaket) {
		supermaket.products.push(this);
	
		console.log(`Product successful added to the supermaket "${supermaket.name}" !`);
	};
	
	// Method Retrieve details of a product
	getOrderInfos() {
		console.log(`
		Order Infos:
		-----------------
		
		ID: ${this.id}
		Customer ID: [ ${this.customerId} ]
		Product ID: [ ${this.productID} ]
		Quantity: ${this.quantity}
		`);
	};

	//! Allow an order to be canceled, and restore the stock
	cancelOrder(supermaket) {
		supermaket.orders = supermaket.orders.filter(order => order.id !== this.id);
		
		console.log(`This order has just successfully been deleted!`);
	};
	
	//! Update the stock (reduce quantity) when an order is placed.
	processOrder(supermaket) {
		const product = supermaket.products.filter(product => product.id !== this.productId);
		product.quantity -= this.quantity; 
		
		console.log(`Order have been placed succefully !`);
	}
}

export default Order;

// Test
const order = new Order("CustomerID123", "ProductID456", 10);
// order.getOrderInfos();

File: Version-Python/models/order.py
from uuid import uuid4
from datetime import datetime

class Order:
    def __init__(self, customer_id, product_id, quantity):
        if not all([customer_id, product_id, quantity]):
            raise ValueError("All required fields must be provided!")

        self.id = str(uuid4())
        self.customer_id = customer_id
        self.product_id = product_id
        self.quantity = quantity
        self.date = datetime.now()

    def create_order(self, supermarket):
        supermarket.orders.append(self)
        print(f'Order successful added to the supermarket "{supermarket.name}"!')

    def get_order_infos(self):
        print(f"""
        Order Infos:
        -----------------
        
        ID: {self.id}
        Customer ID: [ {self.customer_id} ]
        Product ID: [ {self.product_id} ]
        Quantity: {self.quantity}
        """)

    def cancel_order(self, supermarket):
        supermarket.orders = [o for o in supermarket.orders if o.id != self.id]
        print("This order has just successfully been deleted!")

    def process_order(self, supermarket):
        product = next((p for p in supermarket.products if p.id == self.product_id), None)
        if product:
            product.quantity -= self.quantity
            print("Order have been placed successfully!")
